### PR TITLE
Phase 6: Aft-Foil Loss Upweighting — Stronger Gradient for Wake Nodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The dedicated aft-foil SRF head (PR #2104, merged) provides a learned correction for ID=7 surface nodes **after** the Transolver trunk. However, the trunk itself is still trained with equal loss weight for all surface nodes. The aft foil lives in the wake of the fore foil — its pressure distribution is harder to predict and more sensitive to gap/stagger geometry.

**Idea:** Upweight aft-foil (ID=7) nodes in the main surface L1 loss by a factor of 1.5–2.0×. This forces the trunk's attention mechanism to develop richer, more discriminative representations for aft-foil nodes, directly improving the inputs that flow into the SRF head. The mechanism is complementary to (not redundant with) the dedicated SRF head:

- SRF head: post-hoc residual correction  
- Loss upweighting: changes what the trunk optimizes for

**Why this targets p_tan:** The p_tan gap (30.05 single-model vs 29.1 ensemble) is primarily an aft-foil prediction problem. Extra gradient pressure on aft-foil nodes should close this gap. Low risk — the weight is a continuous dial; at 1.0× it recovers the exact baseline.

**Note:** Prior PR #1893 (Foil-2 Loss Upweighting) was tried before the dedicated SRF head existed — this tests the interaction term: loss upweighting WITH the aft-foil SRF head already in baseline.

## Instructions

**Step 1: Add flag to argparse and Config:**
```python
parser.add_argument('--aft_foil_loss_weight', type=float, default=1.0,
                    help='Loss weight multiplier for aft-foil (ID=7) surface nodes (1.0=baseline)')
```
Add `aft_foil_loss_weight: float = 1.0` to the Config dataclass.

**Step 2: Build the aft-foil node mask.**
The same SAF-based proxy used by `--aft_foil_srf` already exists in the training loop. Find where `_aft_foil_mask` is computed (grep for `aft_foil_mask` in train.py) and reuse it. If the mask isn't available at loss-computation time, reconstruct using the same threshold (SAF distance > 0.005, is_tandem, is_surface).

**Step 3: Apply weighted mean in the surface loss computation.**
Find where `surf_loss` is computed (grep for `surf_loss` or surface loss in the training loop). Modify:

```python
if cfg.aft_foil_loss_weight != 1.0 and _aft_foil_mask is not None:
    node_weights = torch.ones(B, N, device=device)
    node_weights[_aft_foil_mask] = cfg.aft_foil_loss_weight
    # Weighted mean: divide by sum of weights, not node count
    weighted_errs = abs_err * node_weights.unsqueeze(-1)  # [B, N, C]
    surf_loss = weighted_errs[surface_mask].sum() / node_weights.unsqueeze(-1)[surface_mask].sum()
else:
    surf_loss = abs_err[surface_mask].mean()
```

**Note:** Use weighted mean (not weighted sum) to keep the loss scale comparable to baseline. This prevents the optimizer from seeing an artificially inflated loss scale.

**Step 4: Run sweep across 2 weights × 2 seeds:**

```bash
# Weight 1.5, seed 42 (run this first)
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/aft-foil-lw-1.5-s42" \
  --wandb_group phase6/aft-foil-loss-weight \
  --aft_foil_loss_weight 1.5 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf

# Weight 1.5, seed 73
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/aft-foil-lw-1.5-s73" \
  --wandb_group phase6/aft-foil-loss-weight \
  --aft_foil_loss_weight 1.5 --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf

# If weight=1.5 looks promising, also run weight=2.0 on both seeds
# (same flags, --aft_foil_loss_weight 2.0, --wandb_name "tanjiro/aft-foil-lw-2.0-s42/73")
```

Report all 4 surface MAE metrics plus val/loss for each run. Include the W&B run IDs.

## Baseline

**Current single-model baseline (PR #2104, +aft_foil_srf, 8-seed mean, seeds 42-49):**

| Metric | Baseline mean | Std | Target to beat |
|--------|--------------|-----|----------------|
| p_in   | 13.19        | 0.33 | < 13.19       |
| p_oodc | 7.92         | 0.17 | < 7.92        |
| p_tan  | 30.05        | 0.36 | < 30.05       |
| p_re   | 6.45         | 0.07 | < 6.45        |

**Merge criterion:** 2-seed average must beat the baseline targets on p_tan (primary) and not regress on p_in/p_oodc/p_re.

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro --wandb_name "tanjiro/baseline-check" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Surface Pressure MAE (lower is better)

| Run | p_in | p_oodc | p_tan | p_re | val_loss | W&B |
|-----|------|--------|-------|------|----------|-----|
| **Baseline mean (PR #2104, 8 seeds)** | **13.19** | **7.92** | **30.05** | **6.45** | — | — |
| aft-lw-1.5-s42 | 12.95 | 7.75 | 30.57 | 6.50 | 0.390 | `ghq78pj6` |
| aft-lw-1.5-s73 | 13.80 | 7.77 | 30.53 | 6.36 | 0.393 | `4r0rksil` |
| aft-lw-2.0-s42 | **13.76** | 7.75 | **29.91** | 6.57 | 0.389 | `svqikz02` |
| aft-lw-2.0-s73 | **12.96** | 7.84 | 31.08 | 6.49 | 0.392 | `rk53nwzs` |

**Mean across weight=1.5 seeds:** p_in=13.37, p_oodc=7.76, p_tan=30.55, p_re=6.43
**Mean across weight=2.0 seeds:** p_in=13.36, p_oodc=7.80, p_tan=30.50, p_re=6.53

### Analysis

**Results are mixed — no clear improvement over baseline p_tan (30.05), with high variance between seeds.**

1. **p_tan did not improve on average.** Weight=1.5 mean: 30.55 (+0.5 vs baseline). Weight=2.0 mean: 30.50 (+0.45 vs baseline). Only one run (w=2.0, s42) beat the target at p_tan=29.91, but its paired seed (w=2.0, s73) was worse at 31.08 — within normal seed variance.

2. **p_in was similar to baseline.** Weight=1.5 mean (13.37) and weight=2.0 mean (13.36) are within baseline std (13.19 ± 0.33). No systematic regression.

3. **p_oodc slightly improved.** All 4 runs: 7.75–7.84 vs baseline 7.92. Small but consistent improvement across all seeds and weights.

4. **Seed variance dominates.** The range between seeds is ±0.9 on p_tan, masking any treatment effect. With 2 seeds it's difficult to establish significance.

5. **Interaction with existing `--aft_foil_srf`:** The aft-foil SRF head already provides a dedicated correction for these nodes. Adding loss upweighting on top may not provide additional benefit since the SRF head is already compensating for aft-foil errors. The loss upweighting pushes the trunk to develop better representations for aft-foil nodes, but the SRF head post-hoc correction may absorb the remaining error regardless.

### Commands Used

```bash
# Weight 1.5, seed 42 (GPU 0)
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_group phase6/aft-foil-loss-weight --wandb_name "tanjiro/aft-foil-lw-1.5-s42" \
  --aft_foil_loss_weight 1.5 --seed 42 \
  --aft_foil_srf --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# Weight 2.0 — same flags with --aft_foil_loss_weight 2.0
```

Peak memory: ~52-97 GB (similar to baseline)
All runs: 154-155 epochs, hit 180-min wall-clock timeout.

### Suggested follow-ups

1. **More seeds for w=2.0** — The single best run (w=2.0, s42, p_tan=29.91) suggests a potential signal. With 4 seeds instead of 2, we could better separate true effect from variance.
2. **Loss upweighting without aft_foil_srf** — Test whether loss upweighting alone (no SRF head) provides cleaner signal. The SRF head may be absorbing the benefit of loss upweighting.
3. **Weight=3.0** — If there is a real effect at w=2.0, w=3.0 might amplify it enough to be detectable within 2 seeds.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2121 because ADVISOR: Closing after thorough exploration. Results:

**w=1.5 (2 seeds):** p_in 13.37 | p_oodc 7.76 | p_tan 30.55 | p_re 6.43
**w=2.0 (2 seeds):** p_in 13.36 | p_oodc 7.80 | p_tan 30.50 | p_re 6.53

**Analysis:** The p_oodc improvement is consistent and real across all 4 runs (7.75-7.84 vs 7.92 baseline), confirming that aft-foil loss upweighting does improve the trunk's wake-region representations for OOD-C generalization. However, p_tan (our primary target) regresses by +1.5-1.7% at both weights — the w=2.0/s42 p_tan=29.91 is contradicted by the paired s73 at 31.08, indicating seed variance rather than a real effect.

**Key insight recorded:** Foil-specific loss weighting improves p_oodc but trades off against p_tan. The SRF head may be absorbing most of the trunk-level benefit before it reaches tandem predictions, exactly as you hypothesized.

Good experimental execution and analysis, tanjiro. Reassigning you to a fresh direction..
